Fix/typescript errors and seeders

Escrodri opened the pull request:
# Fix TypeScript Errors, Database Synchronization, Soft Delete Implementation, Automatic end_at Calculation, and Medical Workflow Improvements

## Resumen
Esta PR corrige múltiples errores de TypeScript, sincroniza los modelos con las migraciones de base de datos, corrige los seeders, actualiza completamente la documentación de Swagger, implementa un sistema completo de borrado lógico (soft delete) para citas médicas con mejoras de seguridad y autorización, agrega cálculo automático de `end_at` con campos opcionales, y mejora significativamente el flujo de trabajo médico con nuevas funcionalidades.

## Problemas Resueltos

### 1. Errores de TypeScript
- **Modelos inconsistentes**: Corregidos nombres de campos entre modelos y migraciones
- **Campos inexistentes**: Eliminados campos que no existían en la base de datos
- **Tipos incorrectos**: Actualizados ENUMs y tipos de datos

### 2. Errores de Base de Datos
- **Foreign key constraints**: Corregidos seeders con IDs hardcodeados
- **Validaciones**: Actualizados campos obligatorios vs opcionales

### 3. Documentación Swagger
- **Endpoints duplicados**: Eliminada duplicación de rutas
- **Prefijos incorrectos**: Agregado prefijo `/api` a todas las rutas
- **Esquemas desactualizados**: Sincronizados con cambios de modelos
- **Rutas inconsistentes**: Corregidas URLs entre documentación y rutas reales

### 4. Sistema de Borrado Lógico (Soft Delete)
- **Auditoría médica**: Implementado borrado lógico para cumplimiento legal
- **Trazabilidad**: Historial completo de citas preservado
- **Recuperación**: Posibilidad de restaurar citas canceladas
- **Estados de cita**: Sistema completo de estados (scheduled, completed, cancelled, no_show)

### 5. Seguridad y Autorización
- **Middlewares de autorización**: Validación de ownership para operaciones CRUD
- **Seguridad granular**: Solo el dueño puede modificar/eliminar sus citas
- **Validación robusta**: Manejo de errores y validaciones mejoradas
- **Admin simplificado**: Solo password embebido, sin autenticación de usuario

### 6. Mejoras en Flujo Médico
- **Síntomas opcionales**: Los síntomas se agregan durante la consulta médica
- **Diagnósticos flexibles**: Los médicos pueden agregar diagnósticos al final
- **Médicos editando pacientes**: Los médicos pueden actualizar citas de pacientes
- **Buenas prácticas REST**: DELETE con query parameters en lugar de body

## Cambios Principales

### Modelos Corregidos
- **`Medic.ts`**: 
  - `specialty` → `speciality`
  - `licence_num` → `license_num`
  - Eliminado `schedule_from`
  - Actualizado ENUM de especialidades
- **`Appointment.ts`**:
  - Reintroducido campo `end_at` (calculado automáticamente como `start_at + 30 minutos` si no se proporciona)
  - `symptoms` **OPCIONAL** (se agregan durante la consulta médica)
  - `diagnostic` opcional (lo completa el médico al final)
  - Nuevos campos: `status`, `deleted_at`, `cancellation_reason`
  - Métodos de borrado lógico: `softDelete()`, `restore()`, `findActive()`, `findDeleted()`

### Servicios Actualizados
- **`user.service.ts`**: Corregidos nombres de campos en operaciones CRUD
- **`appointment.service.ts`**: 
  - Cálculo automático de `end_at` como `start_at + 30 minutos` si no se proporciona
  - Nuevos métodos: `cancelAppointment()`, `restoreAppointment()`, `getCancelledAppointments()`
  - Filtros automáticos para excluir citas canceladas
  - Método `permanentDeleteAppointment()` para administradores
  - **Validación robusta de endAt**: Uso de `endAt!` para evitar undefined

### DTOs y Validadores
- **`registerMedic.dto.ts`**: Actualizado ENUM de especialidades
- **`registerMedic.validator.ts`**: Corregidos nombres de campos
- **`updateMedic.validator.ts`**: Sincronizado con modelo
- **`createAppointment.dto.ts`**: `end_at` opcional, `symptoms` **OPCIONAL**, `diagnostic` opcional
- **`createAppointment.validator.ts`**: Validación actualizada - `symptoms` ahora opcional

### Middlewares Mejorados
- **`admin.middleware.ts`**: 
  - **SIMPLIFICADO**: Solo verifica password admin embebido
  - Eliminada verificación de usuario autenticado
  - Eliminada verificación de rol de médico
- **`appointmentAuthorization.middleware.ts`**: 
  - **NUEVO**: Middleware híbrido para médicos y pacientes
  - Médicos pueden actualizar citas de pacientes
  - Pacientes solo pueden actualizar sus propias citas

### Seeders Corregidos
- **IDs dinámicos**: Reemplazados IDs hardcodeados con consultas dinámicas
- **Campos corregidos**: Sincronizados con esquemas de migración
- **Datos válidos**: Corregidos valores de ENUMs
- **Borrado lógico**: Agregados campos `status`, `deleted_at`, `cancellation_reason`
- **Datos realistas**: Estados de citas (completed, scheduled, cancelled, no_show)
- **Integridad referencial**: Solo citas activas en teleconsultations y notifications

### Documentación Swagger
- **28+ endpoints documentados** completamente
- **Esquemas actualizados** con campos de borrado lógico
- **Rutas consistentes** con prefijo `/api`
- **Sin duplicación** de endpoints
- **URLs corregidas**: `/api/paciente/` en lugar de `/api/patient/`
- **Nuevos esquemas**: `CancelAppointmentRequest` para motivos de cancelación
- **`CreateAppointmentRequest`**: Actualizado con `end_at` opcional y `symptoms` opcional
- **`diagnostic`**: Marcado como opcional con descripción clara
- **Query parameters**: Documentados para DELETE con `reason`

### Nuevas Funcionalidades
- **Sistema de borrado lógico**: Cancelación con motivo opcional
- **Restauración de citas**: Posibilidad de reactivar citas canceladas
- **Historial de cancelaciones**: Endpoints para ver citas canceladas
- **Disponibilidad mejorada**: Solo considera citas activas y programadas
- **Middlewares de autorización**: Validación de ownership granular
- **Cálculo automático de `end_at`**: Se calcula como `start_at + 30 minutos` si no se proporciona
- **`diagnostic` opcional**: El médico lo completa durante o después de la consulta
- **`symptoms` opcional**: Se agregan durante la consulta médica
- **Médicos editando pacientes**: Nueva ruta `/api/medic/patient-appointments`
- **Buenas prácticas REST**: DELETE con query parameters

## Archivos Modificados
```
src/models/Medic.ts
src/models/Appointment.ts
src/services/user.service.ts
src/services/appointment.service.ts
src/controllers/appointment.controller.ts
src/controllers/user.controller.ts
src/routes/appointment.routes.ts
src/routes/auth.routes.ts
src/middlewares/auth/admin.middleware.ts ⭐ SIMPLIFICADO
src/middlewares/appointment/appointmentAuthorization.middleware.ts ⭐ MEJORADO
src/dto/user/registerMedic.dto.ts
src/dto/appointment/createAppointment.dto.ts ⭐ SYMPTOMS OPCIONAL
src/resources/medic/medicResource.resource.ts
src/validators/auth/registerMedic.validator.ts
src/validators/user/updateMedic.validator.ts
src/validators/appointment/createAppointment.validator.ts ⭐ SYMPTOMS OPCIONAL
src/seeders/20250115000004-demo-patients.js
src/seeders/20250115000005-demo-medics.js
src/seeders/20250115000006-demo-appointments.js
src/seeders/20250115000007-demo-medical-records.js
src/seeders/20250115000008-demo-teleconsultations.js
src/seeders/20250115000009-demo-notifications.js
src/config/swagger.config.ts
src/migrations/20251027185015-add-end-at-to-appointments.js
src/migrations/20251027191311-add-soft-delete-to-appointments.js (NUEVO)
```

## Resultados
- Servidor inicia sin errores de TypeScript
- Base de datos sincronizada con modelos
- Seeders ejecutan correctamente sin errores
- Documentación Swagger completa y precisa
- API consistente con rutas en español
- Sin duplicación de endpoints
- **Sistema de borrado lógico completamente funcional**
- **Seguridad y autorización implementada**
- **Auditoría médica cumplida**
- **Trazabilidad completa de citas**
- **Flujo médico mejorado**: Síntomas y diagnósticos flexibles
- **Médicos pueden editar citas de pacientes**
- **Buenas prácticas REST implementadas**

## Testing
- Migraciones ejecutadas exitosamente
- Seeders ejecutados sin errores
- Servidor inicia correctamente
- Documentación Swagger accesible
- **Borrado lógico probado**: Cancelación y restauración funcionando
- **Autorización verificada**: Solo owners pueden modificar sus citas
- **Estados de cita**: Todos los estados (scheduled, completed, cancelled, no_show) funcionando
- **Integridad referencial**: Teleconsultations y notifications solo para citas activas
- **Admin middleware**: Solo password requerido
- **Médicos editando pacientes**: Funcionalidad verificada
- **Síntomas opcionales**: Validación actualizada

## Notas Importantes
- **Breaking Changes**: Los campos `specialty` → `speciality` y `licence_num` → `license_num` requieren actualización del frontend
- **Campos eliminados**: `schedule_from` ya no está disponible
- **Campos reintroducidos**: `end_at` ahora es opcional y se calcula automáticamente
- **Campos opcionales**: `symptoms` ahora es opcional (se agregan durante consulta)
- **Campos opcionales**: `diagnostic` es opcional (lo completa el médico)
- **Nuevos campos**: `status`, `deleted_at`, `cancellation_reason` agregados para borrado lógico
- **Cálculo automático**: `end_at` se calcula como `start_at + 30 minutos` si no se proporciona
- **Cumplimiento legal**: Sistema de auditoría médica implementado
- **Seguridad**: Middlewares de autorización requieren autenticación previa
- **Admin simplificado**: Solo requiere password embebido
- **Flujo médico**: Síntomas y diagnósticos se agregan durante la consulta
- **Buenas prácticas**: DELETE usa query parameters

## Endpoints Documentados

### Authentication
- `POST /api/auth/login` - Iniciar sesión
- `POST /api/auth/register/patient` - Registro de pacientes
- `POST /api/auth/register/medic` - Registro de médicos (solo admin con password)
- `GET /api/auth/user` - Obtener usuario actual
- `POST /api/auth/logout` - Cerrar sesión

### Users
- `GET /api/medics/summary` - Resumen de médicos disponibles
- `GET /api/patients/{id}` - Obtener paciente por ID
- `PATCH /api/patients` - Actualizar información de paciente
- `GET /api/medics/{id}` - Obtener médico por ID
- `PATCH /api/medics` - Actualizar información de médico

### Appointments
- `POST /api/medic/appointments` - Crear cita como médico
- `GET /api/medic/appointments` - Listar citas del médico
- `PATCH /api/medic/appointments` - Actualizar cita como médico
- `GET /api/medic/appointments/{id}` - Obtener cita específica del médico
- `DELETE /api/medic/appointments/{id}?reason=...` - Cancelar cita como médico (borrado lógico) ⭐ QUERY PARAMS
- `PATCH /api/medic/appointments/{id}/restore` - Restaurar cita cancelada (médico) ⭐ NUEVO
- `GET /api/medic/appointments/cancelled` - Ver citas canceladas (médico) ⭐ NUEVO
- `POST /api/paciente/appointments` - Crear cita como paciente
- `GET /api/paciente/appointments` - Listar citas del paciente
- `PATCH /api/paciente/appointments` - Actualizar cita como paciente
- `GET /api/paciente/appointments/{id}` - Obtener cita específica del paciente
- `DELETE /api/paciente/appointments/{id}?reason=...` - Cancelar cita como paciente (borrado lógico) ⭐ QUERY PARAMS
- `PATCH /api/paciente/appointments/{id}/restore` - Restaurar cita cancelada (paciente) ⭐ NUEVO
- `GET /api/paciente/appointments/cancelled` - Ver citas canceladas (paciente) ⭐ NUEVO
- `GET /api/appointments/availability` - Verificar disponibilidad de citas ⭐ NUEVO

### Medical Records
- `GET /api/medical_records/{patient_id}` - Obtener historial médico

### 🆕 Nuevas Rutas Médicas
- `PATCH /api/medic/patient-appointments` - **Médicos editando citas de pacientes** ⭐ NUEVO

---

## Rutas Nuevas Agregadas

### 📌 Nuevas Rutas de Soft Delete (6)
1. **`PATCH /api/medic/appointments/{id}/restore`** - Restaurar cita cancelada (médico)
2. **`GET /api/medic/appointments/cancelled`** - Ver citas canceladas (médico)
3. **`PATCH /api/paciente/appointments/{id}/restore`** - Restaurar cita cancelada (paciente)
4. **`GET /api/paciente/appointments/cancelled`** - Ver citas canceladas (paciente)
5. **`GET /api/appointments/availability`** - Verificar disponibilidad de citas (con parámetros completos)
6. **`PATCH /api/medic/patient-appointments`** - **Médicos editando citas de pacientes** ⭐ NUEVO

### 🔄 Rutas Modificadas (2)
- **`DELETE /api/medic/appointments/{id}`**: Ahora usa **query parameters** para `reason` (buenas prácticas REST)
- **`DELETE /api/paciente/appointments/{id}`**: Ahora usa **query parameters** para `reason` (buenas prácticas REST)

## Nuevas Funcionalidades Implementadas

### 🏥 Sistema de Borrado Lógico
- **Cancelación con motivo**: Opcional, registrado en `cancellation_reason`
- **Restauración**: Posibilidad de reactivar citas canceladas
- **Estados**: `scheduled`, `completed`, `cancelled`, `no_show`
- **Auditoría**: Fecha de cancelación en `deleted_at`

### 🔒 Seguridad y Autorización
- **Middlewares de ownership**: Solo el dueño puede modificar/eliminar
- **Validación granular**: Médicos y pacientes tienen permisos específicos
- **Admin simplificado**: Solo password embebido requerido
- **Médicos editando pacientes**: Acceso controlado para gestión médica

### 🩺 Flujo Médico Mejorado
- **Síntomas opcionales**: Se agregan durante la consulta médica
- **Diagnósticos flexibles**: Los médicos los completan al final
- **Médicos editando pacientes**: Pueden actualizar citas para agregar síntomas/diagnósticos
- **Cálculo automático**: `end_at` se calcula automáticamente
- **Estados de consulta**: Flujo natural de trabajo médico

### 📊 Beneficios del Sistema
- **Cumplimiento legal**: Historial médico completo preservado
- **Reportes precisos**: Estadísticas de cancelaciones vs completadas
- **Flexibilidad operativa**: Restauración de citas canceladas por error
- **Trazabilidad completa**: Seguimiento de todos los cambios
- **Performance optimizada**: Consultas eficientes con índices
- **Flujo médico natural**: Síntomas y diagnósticos en el momento correcto
- **Buenas prácticas**: API RESTful con métodos HTTP apropiados

## Comentarios de Code Review Resueltos

### ✅ Todos los comentarios de FacuNeutral implementados:
1. **DELETE con query parameters** - Implementado
2. **Symptoms opcional** - Implementado en DTO, modelo y validador
3. **Admin middleware simplificado** - Solo password embebido
4. **Symptoms y diagnostic opcionales** - Implementado en modelo
5. **Médicos editando pacientes** - Nueva funcionalidad implementada
6. **Validación de endAt** - Corregida con `endAt!`
7. **Validador de síntomas** - Actualizado a opcional
8. **Documentación de cambios** - Implementada (eliminada por solicitud)

### 🎯 Mejoras adicionales implementadas:
- **Documentación Swagger completa** y consistente
- **Rutas corregidas** para coincidir con URLs reales
- **Buenas prácticas REST** aplicadas
- **Código limpio** y bien documentado
- **Flujo médico optimizado** para mejor UX

Escrodri committed:
Fix TypeScript errors and seeders synchronization

Escrodri committed:
Update Swagger documentation

Escrodri committed:
Fix Swagger API paths - add /api prefix

Escrodri committed:
Fix Swagger duplicate endpoints issue

Escrodri committed:
feat: implement soft delete system and fix TypeScript errors

- Add soft delete functionality for appointments with status, deleted_at, cancellation_reason
- Add new endpoints: restore appointments, get cancelled appointments, availability check
- Fix TypeScript errors in models, services, DTOs and validators
- Update Swagger documentation with all new endpoints and parameters
- Add authorization middlewares for appointment ownership validation
- Fix database seeders with dynamic IDs and soft delete fields
- Add migrations for end_at and soft delete columns
- Refactor editUser method and admin middleware for better security

Escrodri committed:
feat: add automatic end_at calculation and make diagnostic optional

- end_at is now optional and automatically calculated as start_at + 30 minutes if not provided
- diagnostic is optional when creating appointments, filled by doctor during/after consultation
- symptoms is required when creating appointments
- Updated DTO, validator, model and Swagger documentation

Escrodri committed:
Merge branch 'develop' into fix/typescript-errors-and-seeders

Conflicts resolved:
- demo-appointments.js: Merged end_at, soft delete fields and updated symptoms/diagnostic from develop
- demo-teleconsultations.js: Kept soft delete filter (deleted_at IS NULL)
- demo-notifications.js: Kept soft delete filter and corrected appointment index

FacuNeutral commented on backend/src/controllers/appointment.controller.ts:
```diff
@@ -697,16 +696,18 @@ export class AppointmentController {
   ) => {
     try {
       const { id } = request.params;
+      const { reason } = request.body;
```
no es buena practica y a veces se elimina el body en peticiones de tipo "delete". Mejor por queries

FacuNeutral commented on backend/src/dto/appointment/createAppointment.dto.ts:
```diff
@@ -2,8 +2,9 @@ export interface CreateAppointmentDto {
   type: "in_person" | "virtual";
   location?: string;
-  symptoms?: string;
+  symptoms: string; // Requerido: el paciente debe describir sus síntomas
```
no estaría bien declarar el sintoma primero, ya que el sintoma se declara al terminar la reunión por el médico

FacuNeutral commented on backend/src/middlewares/auth/admin.middleware.ts:
```diff
@@ -1,20 +1,55 @@
-  if (!isAdmin) {
-    return res.status(403).json({
+    // Verificar que sea médico
+    if (user.rol !== "medico") {
```
ojo, no modificaría esto, no se puede verificar un rol si no existe un usuario logeado, por eso se una password enbebido en el body como solución rápida.

FacuNeutral commented on backend/src/models/Appointment.ts:
```diff
@@ -7,18 +7,21 @@ interface AppointmentAttributes {
-  symptoms?: string;
-  diagnostic?: string;
+  symptoms: string;
+  diagnostic: string;
```
se dan al final de la cita 
```suggestion
  symptoms?: string;
  diagnostic?: string;
```

FacuNeutral commented on backend/src/models/Appointment.ts:
```diff
@@ -7,18 +7,21 @@ interface AppointmentAttributes {
+  diagnostic: string;
   type: "in_person" | "virtual";
   location?: string;
+  status: "scheduled" | "completed" | "cancelled" | "no_show";
+  deleted_at?: Date;
+  cancellation_reason?: string;
```
cualquier cambio de propiedades en la base de datos recomiendo avisar por escrito en sección anuncios

FacuNeutral commented on backend/src/routes/appointment.routes.ts:
```diff
@@ -79,18 +100,33 @@ router.get(
   "/paciente/appointments",
   updateAppointmentValidator,
   handleValidationErrors,
+  canUpdatePatientAppointment,
```
Quizás los médicos necesiten modificar datos de los pacientes

FacuNeutral commented on backend/src/services/appointment.service.ts:
```diff
@@ -14,18 +14,30 @@ export class AppointmentService {
+        const startAtDate = new Date(appointmentData.start_at);
+        const endAtDate = new Date(startAtDate);
+        endAtDate.setMinutes(endAtDate.getMinutes() + 30);
+        endAt = endAtDate.toISOString();
```
falto asegurar que el endAt existe

FacuNeutral commented on backend/src/validators/appointment/createAppointment.validator.ts:
```diff
@@ -49,12 +50,20 @@ export const createAppointmentValidator: ValidationChain[] = [
   body("symptoms")
-    .optional()
+    .notEmpty()
+    .withMessage("Los síntomas son requeridos")
```
```suggestion
```

FacuNeutral commented on backend/.gitignore:
```diff
@@ -1,2 +1,3 @@
-/.env
\ No newline at end of file
+/.env
+/dist
```
Buena observación!

Escrodri committed:
Merge remote-tracking branch 'origin/develop' into fix/typescript-errors-and-seeders

Escrodri committed:
fix: implement code review feedback and improve medical workflow

- Fix DELETE endpoints to use query parameters instead of body (REST best practices)
- Make symptoms optional in DTO, model and validator (added during consultation)
- Simplify admin middleware to use only embedded password
- Allow doctors to update patient appointments (add symptoms, diagnostics)
- Fix endAt validation to prevent undefined errors
- Update Swagger documentation with correct URLs and parameters
- Improve medical workflow: symptoms and diagnostics added during consultation
- Maintain security: patients can only edit their own appointments
- All FacuNeutral code review comments resolved

FacuNeutral committed:
Remove permission checks for appointment access

Removed permission checks for medic and patient when viewing appointments.
